File: python-algorithm/algo_05_dynamic_plan/dp_9_stock/L122.py
#--coding:utf-8--

# 同牛客134
# 题目要求不能同时参与多笔交易
class Solution:
    # #方法1 DP
    def maxProfit(self, prices: List[int]) -> int:
        dp = [[0, 0] for _ in range(len(prices))]
        dp[0][0] = 0
        dp[0][1] = -prices[0]
        for i in range(1, len(prices)):
            dp[i][0] = max(dp[i-1][0], dp[i-1][1]+prices[i])
            dp[i][1] = max(dp[i-1][0]-prices[i], dp[i-1][1])
        return dp[len(prices)-1][0]

    # #方法2 贪心
    # def maxProfit(self, prices: List[int]) -> int:
    #     res = 0
    #     for i in range(1, len(prices)):
    #         res += max(prices[i]-prices[i-1], 0)
    #     return res

File: python-algorithm/algo_05_dynamic_plan/dp_9_stock/L122.cpp


//买卖股票的最佳时机 II
//给定一个数组，它的第 i 个元素是一支给定股票第 i 天的价格。
//设计一个算法来计算你所能获取的最大利润。你可以尽可能地完成更多的交易（多次买卖一支股票）。
//注意：你不能同时参与多笔交易（你必须在再次购买前出售掉之前的股票）。

//贪心算法：1 只要明天价格大于当天价格，当天就买；2 当天买第二天就卖

#include <iostream>
#include <vector>

using namespace std;
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit = 0;
        for(int i = 0; i + 1 < prices.size(); ++i)
            profit += max(prices[i + 1] - prices[i], 0);
        return profit;
    }
};



